tensorflow/contrib/seq2seq/python/ops/layers.py

```diff
@@ -20,15 +20,185 @@
+          prev = tf.matmul(state, weight) + bias
+          return tf.gather(embeddings, tf.argmax(prev, 1))
+        return dec_fn
+    encoder_projection: (optional) given that the encoder might have a
+      different size than the decoder, we project the intial state as
+      described in Bahdanau, 2014 (https://arxiv.org/abs/1409.0473).
+      The optional `encoder_projection` is a
+      `tf.contrib.layers.fully_connected` with
+      `activation_fn=tf.python.ops.nn.tanh`.
+    parallel_iterations: (Default: 32).  The number of iterations to run in
+      parallel.  Those operations which do not have any temporal dependency
+      and can be run in parallel, will be.  This parameter trades off
+      time for space.  Values >> 1 use more memory but take less time,
+      while smaller values use less memory but computations take longer.
+    swap_memory: Transparently swap the tensors produced in forward inference
+      but needed for back prop from GPU to CPU.  This allows training RNNs
+      which would typically not fit on a single GPU, with very minimal (or no)
+      performance penalty.
+    time_major: The shape format of the `inputs` and `outputs` Tensors.
+      If true, these `Tensors` must be shaped `[max_time, batch_size, depth]`.
+      If false, these `Tensors` must be shaped `[batch_size, max_time, depth]`.
+      Using `time_major = True` is a bit more efficient because it avoids
+      transposes at the beginning and end of the RNN calculation.  However,
+      most TensorFlow data is batch-major, so by default this function
+      accepts input and emits output in batch-major form.
+    scope: VariableScope for the created subgraph; defaults to "RNN".
+  Returns:
+    A pair (outputs, state) where:
+      outputs: The RNN output `Tensor`.
+        If time_major == False (default), this will be a `Tensor` shaped:
+          `[batch_size, max_time, cell.output_size]`.
+        If time_major == True, this will be a `Tensor` shaped:
+          `[max_time, batch_size, cell.output_size]`.
+      state: The final state.  If `cell.state_size` is an int, this
+        will be shaped `[batch_size, cell.state_size]`.  If it is a
+        `TensorShape`, this will be shaped `[batch_size] + cell.state_size`.
+  Raises:
+    TypeError: If `cell` is not an instance of RNNCell.
+  """
+  with vs.variable_scope(scope or "decoder") as varscope:
+    # Project initial_state as described in Bahdanau et al. 2014
+    # https://arxiv.org/abs/1409.0473
+    state = encoder_projection(initial_state, cell.output_size)
+    # Setup of RNN (dimensions, sizes, length, initial state, dtype)
+    # Setup dtype
+    dtype = state.dtype
+    if not time_major:
+      # [batch, seq, features] -> [seq, batch, features]
+      decoder_inputs = array_ops.transpose(decoder_inputs, perm=[1, 0, 2])
+    # Get data input information
+    batch_size = array_ops.shape(decoder_inputs)[1]
+    decoder_input_depth = int(decoder_inputs.get_shape()[2])
```

Comment:
make sure that decoder_inputs.get_shape()[2].value is not None;  if it is, raise a ValueError.